include/deal.II/matrix_free/vector_data_exchange.h

```diff
@@ -247,6 +247,307 @@ namespace internal
+
+        /**
+         * Pointer into the receive buffer.
+         */
+        std::vector<std::pair<types::global_dof_index, types::global_dof_index>>
+          recv_remote_ptr;
+
+        /**
+         * TODO
+         */
+        std::vector<unsigned int> shifts;
+
+        /**
+         * TODO
+         */
+        std::vector<unsigned int> shifts_ptr;
+
+        /**
+         * Shared-memory ranks from where data is copied.
+         */
+        std::vector<unsigned int> recv_sm_ranks;
+
+        /**
+         * Pointer into the receive buffer.
+         */
+        std::vector<unsigned int> recv_sm_ptr;
+
+        /**
+         * Indices to be copied from shared-memory neighbor.
+         */
+        std::vector<unsigned int> recv_sm_indices;
+
+        /**
+         * The same as above, but, compressed: number of entries to be copied.
+         */
+        std::vector<unsigned int> recv_sm_len;
+
+        /**
+         * The same as above, but, compressed: offset.
+         */
+        std::vector<unsigned int> recv_sm_offset;
+
+        /**
+         * Remote ranks to where data is sent.
+         */
+        std::vector<unsigned int> send_remote_ranks;
+
+        /**
+         * Pointer into the send buffer.
+         */
+        std::vector<unsigned int> send_remote_ptr;
+
+        /**
+         * Indices to be sent.
+         */
+        std::vector<unsigned int> send_remote_indices;
+
+        /**
+         * The same as above, but, compressed: number of entries to be copied.
+         */
+        std::vector<unsigned int> send_remote_len;
+
+        /**
+         * The same as above, but, compressed: offset.
+         */
+        std::vector<unsigned int> send_remote_offset;
+
+        /**
+         * Shared-memory ranks from where data is copied during compress.
+         */
+        std::vector<unsigned int> send_sm_ranks;
+
+        /**
+         * Pointer into buffer.
+         */
+        std::vector<unsigned int> send_sm_ptr;
+
+        /**
+         * Indices to be read during compress.
+         */
+        std::vector<unsigned int> send_sm_indices;
+
+        /**
+         * The same as above, but, compressed: number of entries to be copied.
+         */
+        std::vector<unsigned int> send_sm_len;
+
+        /**
+         * The same as above, but, compressed: offset.
+         */
+        std::vector<unsigned int> send_sm_offset;
+
+        /**
+         * TODO
+         */
+        std::vector<unsigned int> recv_sm_indices_;
+
+        /**
+         * TODO
+         */
+        std::vector<unsigned int> recv_sm_ptr_;
+
+        /**
+         * TODO
+         */
+        std::vector<unsigned int> recv_sm_len_;
+
+        /**
+         * TODO
+         */
+        std::vector<unsigned int> send_sm_indices_;
+
+        /**
+         * TODO
+         */
+        std::vector<unsigned int> send_sm_ptr_;
+
+        /**
+         * TODO
+         */
+        std::vector<unsigned int> send_sm_len_;
```

Comment:
I have not arrived yet at the implementation, but I find the names here that just differ by the final underscore from other variables too close in the name. Can you find more appropriate names (here or for the variables above).